### PR TITLE
rfcs: graph api: support Dropout operation

### PR DESCRIPTION
**Pull request by ElaineBao**
# Description

The RFC proposes adding a Dropout operation in oneDNN Graph.

Rendered version: [link](https://github.com/uxlfoundation/oneDNN/blob/yixin/rfc/dropout/rfcs/20250910-graph-api-dropout/README.md)

### CODE REVIEW

File: rfcs/20250910-graph-api-dropout/README.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+- Dropout will use the Philox algorithm for random number generation.
+- Data types for `src`, `p`, `dst` and `mask` are consistent with the
+  current oneDNN primitive design.
+- The `mask` output is optional, allowing flexibility for scenarios where
+  the mask is not needed.
+- The `offset` input is introduced to ensure reproducibility and correct
+  partitioning across devices and threads, following best practices from PyTorch
+  and cuDNN.
+- Data types for `seed` and `offset` are aligned to PyTorch to ensure
```
Comment (TaoLv) [inline rfcs/20250910-graph-api-dropout/README.md]: How about the data types of probability and mask in PyTorch?

File: rfcs/20250910-graph-api-dropout/README.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+  compatibility and support large models.
+- A dedicated DropoutBackward operation is not defined, as the backward
```
Comment (TaoLv) [inline rfcs/20250910-graph-api-dropout/README.md]: "not defined" or "not needed"?

File: rfcs/20250910-graph-api-dropout/README.md [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+```cpp
+
+using namespace dnnl::graph;
+
+graph g = graph(engine::kind::gpu);
+
+logical_tensor src = logical_tensor(ID_SRC, data_type::f16, {256, 1000});
+logical_tensor seed = logical_tensor(ID_SEED, data_type::u64, {1});
+logical_tensor offset = logical_tensor(ID_SEED, data_type::u64, {1});
```
Comment (TaoLv) [inline rfcs/20250910-graph-api-dropout/README.md]:
```suggestion
logical_tensor offset = logical_tensor(ID_OFFSET, data_type::u64, {1});
```

File: rfcs/20250910-graph-api-dropout/README.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+**Operation Kind:**
+
+- `Dropout` (C++), `dnnl_graph_op_dropout` (C)
+
+**Inputs:**
+
+- `src`: input tensor (`f32` / `bf16` / `f16`)
+- `seed`: random seed (`uint64_t`, single value)
+- `offset`: random offset (`uint64_t`, single value)
+- `p`: dropout probability (`f32`, single value)
```
Comment (TaoLv) [inline rfcs/20250910-graph-api-dropout/README.md:106]: Does that mean seed, offset, and p will be changed among different iterations?

File: rfcs/20250910-graph-api-dropout/README.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+dropout.add_input(p);
+dropout.add_output(dst);
+dropout.add_output(mask);
+
+g.add_op(dropout);
+g.finalize();
+
+```
+
+**Open Question:**
```
Comment (TaoLv) [inline rfcs/20250910-graph-api-dropout/README.md:170]: Random place: please consider and document how it will be tested, especially about the random-ness part.

File: rfcs/20250910-graph-api-dropout/README.md (hunk 2 above)
Comment (ElaineBao) [reply rfcs/20250910-graph-api-dropout/README.md]: I think it's not defined. We always have an option to define a dedicated DropoutBackward operation to make it more clear that forward uses Dropout, backward uses DropoutBackward, although underneath the math formula is the same.

File: rfcs/20250910-graph-api-dropout/README.md (hunk 1 above)
Comment (ElaineBao) [reply rfcs/20250910-graph-api-dropout/README.md]: probability is float32,
for mask, the aten sdpa doesn't define an output of mask. the aten dropout has this output, it supports u8 and boolean dtype of mask.

File: rfcs/20250910-graph-api-dropout/README.md (hunk 4 above)
Comment (ElaineBao) [reply rfcs/20250910-graph-api-dropout/README.md:106]: p and seed most likely will not change among different iterations, only offset will change

File: rfcs/20250910-graph-api-dropout/README.md (hunk 5 above)
Comment (ElaineBao) [reply rfcs/20250910-graph-api-dropout/README.md:170]: Thanks for the suggestion, will do.

File: rfcs/20250910-graph-api-dropout/README.md (hunk 4 above)
Comment (TaoLv) [reply rfcs/20250910-graph-api-dropout/README.md:106]: Then why do we put p and seed as inputs, not attributes? Are they on device memory?

File: rfcs/20250910-graph-api-dropout/README.md (hunk 1 above)
Comment (TaoLv) [reply rfcs/20250910-graph-api-dropout/README.md]: Since you explicitly mentioned that `Data types for seed and offset are aligned to PyTorch`. It makes me wondering if probability and mask are not aligned.

> the aten dropout has this output, it supports u8 and boolean dtype of mask.

Better to mention this in the RFC as well and why we choose `u8`, not `boolean`. What if we are asked to support `boolean` mask?

File: rfcs/20250910-graph-api-dropout/README.md (hunk 2 above)
Comment (TaoLv) [reply rfcs/20250910-graph-api-dropout/README.md]: Thank you for the explain. It still sounds like "not needed". Probably we will not need to define two operations to implement the same math formula. Just like we don't define backward operation for MatMul.

File: rfcs/20250910-graph-api-dropout/README.md (hunk 4 above)
Comment (ElaineBao) [reply rfcs/20250910-graph-api-dropout/README.md:106]: seed is still an input in framework side, it can be changed if user wants to, but this use scenario is not common. That's why seed is kept as an input.
for offset, I left it as an open question.

File: rfcs/20250910-graph-api-dropout/README.md (hunk 1 above)
Comment (ElaineBao) [reply rfcs/20250910-graph-api-dropout/README.md]: ok, I'll mention it, thanks for the suggestion. The reason why mask is chosen as u8 is to facilitate integration between the graph API and primitive API, as primitive API already supports u8 mask. We can support boolean if it's requested in the future.

File: rfcs/20250910-graph-api-dropout/README.md (hunk 2 above)
Comment (ElaineBao) [reply rfcs/20250910-graph-api-dropout/README.md]: ok, let me clarify it.

File: rfcs/20250910-graph-api-dropout/README.md [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+## Adoption in Frameworks and Libraries
+
+**PyTorch**
+
+troch.nn.Dropout [[#2]][2] takes `p` (probability, f32) and `inplace` (bool) as attributes,
+with one input and one output. Internally, the random number generator used to
+generate the dropout mask (u8/bool) differs by device:
+
+- CPU: Uses torch.Generator backed by the C++ Mersenne Twister engine [[#3]][3],
+  which consumes a 64-bit seed.
```
Comment (mgouicem) [inline rfcs/20250910-graph-api-dropout/README.md:55]: This is interesting, does it mean that PyTorch deterministic mode is not supported on CPU by default? Or that dropout has to run sequentially?

File: rfcs/20250910-graph-api-dropout/README.md (hunk 5 above)
Comment (mgouicem) [reply rfcs/20250910-graph-api-dropout/README.md:170]: If we stick with Philox, the generation is actually deterministic: for a given output element at index idx, the generated rng value is Philox(seed, offset + idx).

The thing to be careful with is that the flattened `idx` should be consistent between reference and optimized impl (physical offset to base ptr should do).

File: rfcs/20250910-graph-api-dropout/README.md (hunk 4 above)
Comment (mgouicem) [reply rfcs/20250910-graph-api-dropout/README.md:106]: Extra things to consider here wrt input vs attribute: using attribute will also trigger extra jitting and increase cache occupancy. So unless we jit completely different code based on p and seed (which we don't), it might be beneficial to keep those as inputs.

If those are never on device, would it make sense to only support host_side_scalar for those?

File: rfcs/20250910-graph-api-dropout/README.md (hunk 6 above)
Comment (ElaineBao) [reply rfcs/20250910-graph-api-dropout/README.md:55]: Yes,  the random number generator used for Dropout CPU is the default PyTorch CPU RNG, which is non determinitic (evidence: [link](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/CPUGeneratorImpl.cpp#L57)), so If users run dropout (or any random op) on CPU in a single thread and set the seed, they can get deterministic results.
If they use multiple threads, the order of RNG calls can vary, so results may not be deterministic across runs, even with the same seed.

File: rfcs/20250910-graph-api-dropout/README.md (hunk 4 above)
Comment (ElaineBao) [reply rfcs/20250910-graph-api-dropout/README.md:106]: Yes, I think supporting host scalar is enough for seed, offset and p.

**Review by TaoLv — APPROVED** [review]

**Review by vpirogov — APPROVED** [review]
